core/src/test/java/org/infinispan/distribution/MagicKey.java

```diff
@@ -24,57 +33,99 @@
     */
    private static final long serialVersionUID = -835275755945753954L;
 
+   private static final WeakHashMap<ConsistentHash, int[]> hashCodes = new WeakHashMap<>();
+   private static final AtomicLong counter = new AtomicLong();
+
    /**
     * The name is used only for easier debugging and may be null. It is not part of equals()/hashCode().
     */
    private final String name;
    private final int hashcode;
+   /**
+    * As hashcodes can collide, using counter makes the key unique.
+    */
+   private final long unique;
    private final int segment;
    private final String address;
 
    public MagicKey(String name, Cache<?, ?> primaryOwner) {
       this.name = name;
-      address = addressOf(primaryOwner).toString();
-      Random r = new Random();
-      Object dummy;
-      int attemptsLeft = 1000;
-      do {
-         // create a dummy object with this hashcode
-         final int hc = r.nextInt();
-         dummy = new Integer(hc);
-         attemptsLeft--;
-
-      } while (!isFirstOwner(primaryOwner, dummy) && attemptsLeft >= 0);
-
-      if (attemptsLeft < 0) {
-         throw new IllegalStateException("Could not find any key owned by " + primaryOwner);
+      Address primaryAddress = addressOf(primaryOwner);
+      this.address = primaryAddress.toString();
+
+      ConsistentHash ch = primaryOwner.getAdvancedCache().getDistributionManager().getConsistentHash();
+      int segment = findSegment(ch.getNumSegments(), s -> primaryAddress.equals(ch.locatePrimaryOwnerForSegment(s)));
+      if (segment < 0) {
+         throw new IllegalStateException("Could not find any segment owned by " + primaryOwner +
+            ", primary segments: " + segments(primaryOwner));
       }
-      // we have found a hashcode that works!
-      hashcode = dummy.hashCode();
-      segment = primaryOwner.getAdvancedCache().getDistributionManager().getReadConsistentHash().getSegment(this);
+      this.segment = segment;
+      hashcode = getHashCodeForSegment(ch, segment);
```

Comment:
Hmm, it could be changed to `Collection<Integer> findSegments(...)` and `getHashCodeForSegments(ConsistentHash ch, Collection<Integer> segment)` but the latter function would then return first cached hashcode mapping to one of the segments - and that would result in all keys for given owner ending up in the same segment.